### PR TITLE
Add accept_ignored_values to field spec

### PR DESCRIPTION
**Pull request by flash1293**
## What does this PR do?

This PR adds a new property `skip_ignored_fields` to the system test spec which is planned to be used in the `elastic-package` system tests.

## Why is it important?

As discussed in https://github.com/elastic/elastic-package/pull/1738 , it's sometimes important to mute errors during testing because they are not indicative of a real problem but related to the test setup.

This addition allows to override test failures similar to comments muting linting errors in source code.

## Checklist

- [x] I have added test packages to [`test/packages`](https://github.com/elastic/package-spec/tree/main/test/packages) that prove my change is effective.
- [x] I have added an entry in [`spec/changelog.yml`](https://github.com/elastic/package-spec/blob/main/spec/changelog.yml).

### CODE REVIEW

File: test/packages/good/data_stream/foo/fields/some_fields.yml [hunk 1]
```diff
@@ -27,6 +27,11 @@
   description: A field with a pattern defined
   pattern: '^[a-zA-Z]$'
   example: abcd
+- name: foobar
+  type: keyword
+  ignore_above: 5
+  description: A field that can be ignored without failing tests
+  accept_ignored_values: true
```
Comment (ruflin) [inline test/packages/good/data_stream/foo/fields/some_fields.yml:34]: Brainstorming around a better name:

* Something like skip or ignore, not validate
* Something around errors?
* Unfortunately already called ignore 

`skip_ignore_above`, `skip_ignore_above_error`? -> I made it specific as I expect in the future we can tell apart the different reasons. Maybe for now `skip_ignore_error(s)`

Did you used `ignored` on purpose vs `ignore`?

File: spec/integration/data_stream/fields/fields.spec.yml [hunk 2]
```diff
@@ -318,6 +318,16 @@ spec:
         type: integer
         default: 1024
 
+      accept_ignored_values:
+        description: >
+          If set to true, elastic-package system tests will not fail if values
+          for this field can't be indexed for any incoming documents. This
+          should only be used if the failure is related to the test environment
+          and wouldn't happen in production, mitigate the issue via mapping or
+          pipeline changes otherwise. Defaults to false.
```
Comment (jsoriano) [inline spec/integration/data_stream/fields/fields.spec.yml:327]: Are there actual cases where we want to ignore these failures? Or this is a temporary measure till this issue is fixed in packages?

If there are actual cases where we want to ignore these failures, and as described here, this is only expected in tests, I think we should add this to the test config files. We have other test settings like `skip`, `numeric_keyword_fields` or `dynamic_fields` that are used to customize how validations behave on certain cases. I think this setting would fit there.

If this is a temporary measure to help merging https://github.com/elastic/elastic-package/pull/1738 before the issues are solved, we can think on other approaches. Something we have done in some cases is to apply new validations only on packages starting on a specific spec version. On these cases the issue is eventually fixed as packages upgrade their spec versions.

File: test/packages/bad_accept_ignore_above/changelog.yml [hunk 3]
```diff
@@ -0,0 +1,26 @@
+- version: 1.0.0
```
Comment (jsoriano) [inline test/packages/bad_accept_ignore_above/changelog.yml:1]: Test packages need to be referenced in `code/go/pkg/validator/validator_test.go` to be actually used in testing.

**Review by jsoriano — CHANGES_REQUESTED** [review]
If needed, I think we should make this a test config setting rather than part of the field definition.

File: spec/integration/data_stream/fields/fields.spec.yml (hunk 2 above)
Comment (flash1293) [reply spec/integration/data_stream/fields/fields.spec.yml:327]: > Are there actual cases where we want to ignore these failures? Or this is a temporary measure till this issue is fixed in packages?

There is at least one case in spring boot.

The test config files are indeed a better place, didn't think of those - I will adjust.

**Comment by flash1293** [comment]
@ShourieG This was an old version, I just pushed a change, could you have a look? I added it to the system test definition now.

File: code/go/pkg/validator/validator_test.go [hunk 4]
```diff
@@ -49,7 +49,7 @@ func TestValidateFile(t *testing.T) {
 			},
 		},
 		"bad_deploy_variants": {
-			"_dev/deploy/variants.yml",
+			"_dev/deploy/variants.yVml",
```
Comment (ShourieG) [inline code/go/pkg/validator/validator_test.go]: @flash1293 is this a typo here ? Noticed this while going through the change.

**Comment by flash1293** [comment]
>  is there any other way we can include this, example adding it in the validations.yml file

I think it should be a per-field setting, not a wholesale config to ignore all ignored fields in all tests as this would shadow newly introduced issues. Is there a way to do this without releasing a new package spec version?

**Comment by flash1293** [comment]
@ShourieG I might misunderstand, but in order to extend the `validation.yml` to also take a list of fields to not check `_ignored` for, don't we need to release a version of the package spec as well? By my understanding currently `validation.yml` can only completely exclude certain checks, which is not what we want here

**Comment by flash1293** [comment]
Hmm, maybe we could have something like `SVR00006:<field name>` would be an option (but it still seems like a workaround):

```yaml
errors:
  exclude_checks:
   - "SVR00006:error.message" # do not fail on error.message getting ignored
```

What do you think @jsoriano ? To me it seems like the system test config is the right place for this, but I don't have a strong opinion really.

**Comment by jsoriano** [comment]
> What do you think @jsoriano ? To me it seems like the system test config is the right place for this, but I don't have a strong opinion really.

Yeah, I prefer to have it in the test config.

> Having this tied to the package-spec forces us to increase the format_version

The spec for test config files is barely validated, we accept almost anything there (due to this [`additionalProperties: true`](https://github.com/elastic/package-spec/pull/743/files#diff-5b01b03352fa50eee40a9e3b46fc32433be1958ede274d9c00dc7c203c07a1c1R7)), so using this feature won't force to increase the `format_version`, at least as it is now being considered.

File: spec/integration/data_stream/_dev/test/system/config.spec.yml [hunk 5]
```diff
@@ -11,4 +11,15 @@ spec:
...
-      example: 10m
\ No newline at end of file
+      example: 10m
+    skip_ignored_fields:
+      description: >
+        If listed here, elastic-package system tests will not fail if values
+        for the specified field names can't be indexed for any incoming documents. This
+        should only be used if the failure is related to the test environment
+        and wouldn't happen in production. Mitigate the issue via mapping or
+        pipeline changes otherwise.
```
Comment (jsoriano) [inline spec/integration/data_stream/_dev/test/system/config.spec.yml:21]: :+1:

File: test/packages/good/data_stream/foo/_dev/test/system/test-default-config.yml [hunk 6]
```diff
@@ -1,2 +1,4 @@
 wait_for_data_timeout: 10m
+skip_ignored_fields:
+  - error.message
```
Comment (jsoriano) [inline test/packages/good/data_stream/foo/_dev/test/system/test-default-config.yml:3]: Nit. Could we also add a good example in the good_v3 package?

File: code/go/pkg/validator/validator_test.go [hunk 7]
```diff
@@ -132,6 +132,12 @@ func TestValidateFile(t *testing.T) {
 				"field streams.0.vars.3: Must not be present",
 			},
 		},
+		"bad_skip_ignored_fields": {
```
Comment (jsoriano) [inline code/go/pkg/validator/validator_test.go:135]: This package seems to have more issues, did you create it by copying an old package? Consider creating it from scratch with `elastic-package create package` and `elastic-package create data-stream`.

**Review by jsoriano — COMMENTED** [review]
This approach LGTM, let's solve the issues with CI. Thanks!

**Comment by jsoriano** [comment]
As this change is not required for https://github.com/elastic/elastic-package/pull/1738, could you include the implementation for this feature there before merging this?
Thanks!

**Review by jsoriano — APPROVED** [review]
Please consider adding the implementation for this in https://github.com/elastic/elastic-package/pull/1738 before merging :pray: 

Thanks!

**Comment by flash1293** [comment]
https://github.com/elastic/elastic-package/pull/1738 is merged, could you merge @jsoriano ? (I don't have the permission to do so)

**Comment by jsoriano** [comment]
Merged, thanks!

**Pull request merged**